### PR TITLE
permutedMNIST

### PR DESCRIPTION
**Pull request by karangrewal**
This PR mainly concerns the addition of the permutedMNIST dataset for continual learning. For memory efficiency, the MNIST images are not duplicated, and instead the different permutations are held in memory and applied to each image when being loaded.

Other notable changes:

1. In addition to `PermutedMNIST`, there's also a `ContextDependentPermutedMNIST` dataset class which augments the former by additionally returning a binary sparse context vector. This is useful in the case of dendritic networks when task id is provided and not inferred.
2. `compute_task_indices` in the CL experiment class has been slightly modified so that train/test loaders need not yield 2-item tuples (as was previously assumed), but now can yield any number of batch items as long as the target label is the last one.
3. `DendriticMLP` has been augmented with the `dendrite_sparsity` parameter, so users may specify the level of sparsity in dendritic weights. This is useful because in many experiments, forgetting is only avoided when dendrites are fully dense. This change to `DendriticMLP` has been propagated throughout nupic.research.
4. The new project file in the permutedMNIST project folder runs a dendritic MLP on permutedMNIST (with context provided) in an identical setup to the [XdG model](https://www.pnas.org/content/pnas/115/44/E10467.full.pdf).

### CODE REVIEW

**Comment by subutai** [comment]
@akashvelu There are some changes to dendritic_mlp you might want to look at as well - might conflict with your changes.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            )
+
+        self.contexts = ContextDependentPermutedMNIST.contexts
+
+    def __getitem__(self, index):
+        """
+        Returns an (image, context, target) triplet.
+        """
+        img, target = super().__getitem__(index)
+        task_id = index // len(self.data)
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:127]: What about adding a get_task_id(index) method in PermutedMNIST? It would return an int and might be generally useful. Then this line (and also line 78) should use that method.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        return img, self.contexts[task_id, :], target
```
Comment (subutai) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:128]: Maybe also return `task_id` here?

File: projects/dendrites/permutedMNIST/run_dendritic_network.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        num_classes_per_task = math.floor(num_classes / num_tasks)
+
+        task_indices = defaultdict(list)
+        for i in range(num_tasks):
+            for j in range(num_classes_per_task):
+                task_indices[i].extend(class_indices[j + (i * num_classes_per_task)])
+        return task_indices
+
+    def run_iteration(self):
+        # Reset optimizer before training on a new task, just as context-dependent
```
Comment (subutai) [inline projects/dendrites/permutedMNIST/run_dendritic_network.py:75]: Is this ever called in the script? I couldn't see where.

**Review by subutai — APPROVED** [review]
Looks good. Left a couple of comments and suggestions.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:128]: What's the need? If the context vector is provided, then that itself is task information. If we return `task_id`, then it should be eventually used to select the appropriate context vector, which means contexts would be separate from the dataset class. That would make sense, but it still seems better to do it this way.

Comment (subutai) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:128]: Ok, makes sense.

File: projects/dendrites/permutedMNIST/run_dendritic_network.py (hunk 3 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/run_dendritic_network.py]: I removed it, and placed the optimizer reset within the continual learning loop.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 1 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py]: Added.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        `index`, determines which pixel-wise permutation to apply. Target values are
+        also scaled to be unique to each permutation.
+        """
+        img, target = super().__getitem__(index % len(self.data))
+
+        # Determine which task `index` corresponds to
+        task_id = self.get_task_id(index)
+
+        # Apply permutation to `img`
+        img = permute(img, self.permutations[task_id])
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:81]: where is `permute` coming from?

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        ])
+        super().__init__(root=root, train=train, transform=data_transform,
+                         target_transform=target_transform, download=download)
+
+        self.num_tasks = num_tasks
+
+        # Generate num_tasks random permutations - the first one is the identity
+        # permutation (i.e., regular MNIST), represented below as `None`
+        if PermutedMNIST.permutations is None:
+            PermutedMNIST.permutations = [
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:60]: `permutations` is clearly an instance variable, since it is modified in the init method of a class (the init initializes an instance). Why is it defined as a class variable?

Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:60]: `permutations` is clearly an instance variable, since it is modified in the init method of a class (the init initializes an instance). Why is it defined as a class variable?

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    contexts = None
+
+    def __init__(self, num_tasks, dim_context, train, root=".", target_transform=None,
+                 download=False):
+
+        super().__init__(num_tasks, train, root, target_transform, download)
+
+        # Generate context vectors if `contexts` isn't already defined
+        if ContextDependentPermutedMNIST.contexts is None:
+            ContextDependentPermutedMNIST.contexts = generate_context_vectors(
```
Comment (lucasosouza) [inline nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:117]: Same coment applues here, why is `contexts` a class variable

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:129]: I don't think modifying the signature of dataloader getitem is a good alternative. Dataloader is just data and target, and this case your data happens to be not just an image, but an image but a context. It could potentially be a composition of even more things.

Python is not type enforced and type hints are commonly used with unions of types, that flexibility exists by design. So data and target could be of different types. I think the best alternative here is returning a tuple data (img, context) and target, preserving the original signature.

File: projects/dendrites/permutedMNIST/run_dendritic_network.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+from nupic.research.frameworks.vernon import ContinualLearningExperiment, mixins
+
+
+# ------ Experiment class
+class PermutedMNISTExperiment(mixins.RezeroWeights,
+                              ContinualLearningExperiment):
+    pass
+
+
+# ------ Training & evaluation functions
```
Comment (lucasosouza) [inline projects/dendrites/permutedMNIST/run_dendritic_network.py:47]: Are all dendrites experiment still done through this loop? As a suggestion, this could be rewritten in a few lines of code if using Vernon.

**Review by lucasosouza — COMMENTED** [review]
I will mark as comment so you can decide whether to implement or not.

File: nupic/research/frameworks/vernon/experiments/cl_experiment.py [hunk 8]
```diff
@@ -138,7 +138,8 @@ def create_validation_sampler(cls, config, dataset):
     def compute_task_indices(cls, config, dataset):
         # Assume dataloaders are already created
         class_indices = defaultdict(list)
-        for idx, (_, target) in enumerate(dataset):
+        for idx, batch_item in enumerate(dataset):
```
Comment (lucasosouza) [inline nupic/research/frameworks/vernon/experiments/cl_experiment.py:141]: See comment on preserving the dataloader signature. It is a small change and won't really modify things a lot here, just how you unpack.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:129]: I don't think modifying the signature of dataloader getitem is a good alternative. Dataloader is just data and target, and this case your data happens to be not just an image, but an image but a context. It could potentially be a composition of even more things.

Python is not type enforced and type hints are commonly used with unions of types, that flexibility exists by design. So data and target could be of different types. I think the best alternative here is returning a tuple data (img, context) and target, preserving the original signature.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 5 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:60]: Since we instantiate a `PermutedMNIST` object twice (once for training and once for validation), we need the same pixel-wise permutations applied to images for each task. I found that defining `permutations` as a class variable is the easiest way to accomplish this.

One other way to accomplish the same would be to make `permutations` an optional argument to the `__init__` method, and that way when the validation class is instantiated, we can pass in the training class's permutations so that the same pixel-wise permutations are also applied to the validation examples. However, this may get messy since I'm not sure how the configs in Vernon would do this exactly.

Do you have any preferences on which way to go?

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 6 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:117]: Same reason why `permutations` is a class variable.

File: projects/dendrites/permutedMNIST/run_dendritic_network.py (hunk 7 above)
Comment (karangrewal) [reply projects/dendrites/permutedMNIST/run_dendritic_network.py:47]: Yes. I'm planning to merge everything with Vernon soon, but for now, there are slight modifications I needed to make to the training & evaluation functions that's easier done if I just redefine the functions themselves.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 5 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:60]: There are several ways of solving this, but I don't think class attributes is one of them. Even though it might work in one tested used case, that is not what class attributes are for and can cause hidden bugs. 

Training and validation could share the dataloader. If they can't, and dataloader has to share permutations, then composition makes sense - permutations belonging to a higher level class (like Experiments), and passed as an object to dataloader. 

But the easiest way to solve is just seed it. Have a seed be initialized in the higher class and be passed as an argument to the dataloader, no need to pass all permutations. This will always yield the same results: 
```
g = torch.Generator()
g.manual_seed(30)
torch.randperm(100, generator=g)
```

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (lucasosouza) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:129]: @karangrewal just making sure you saw this comment as well

**Comment by lucasosouza** [comment]
@karangrewal please see comment on dataloader getitem signature, I think it went it unnoticed in the review because I added the comment to an existing thread

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py:129]: @lucasosouza noted! I can make this change so that the dataloader yields ((data, context), target) tuples.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 5 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py]: I think this is a good suggestion. I modified the dataset class so that `seed` is now an argument (specified in the config), and that's used to generate the permutations and contexts.

File: nupic/research/frameworks/pytorch/datasets/permuted_mnist.py (hunk 2 above)
Comment (karangrewal) [reply nupic/research/frameworks/pytorch/datasets/permuted_mnist.py]: Fixed.

**Review by lucasosouza — APPROVED** [review]
